src/state-representation.lisp

```diff
@@ -0,0 +1,234 @@
+            (string
+             (make-instance 'posix-shared-memory-allocation :length (expt 2 num-qubits)
+                                                            :name allocation))
+            (t
+             (assert (= (allocation-length allocation) (expt 2 num-qubits)))
+             allocation))))
+    (multiple-value-bind (amplitudes finalizer)
+        (allocate-vector allocation)
+      ;; initialize amplitudes to zero state (efficiently)
+      (setf (aref amplitudes 0) (cflonum 1)) 
+      (let ((state (make-instance 'pure-state :num-qubits num-qubits 
+                                              :amplitudes amplitudes 
+                                              :allocation allocation)))
+        ;; When the state disappears, make sure the shared
+        ;; memory gets deallocated too.
+        (tg:finalize state finalizer)
+        state))))
+
+(defmethod initialize-instance :after ((state pure-state) &key num-qubits &allow-other-keys)
+  (cond 
+    ((and (slot-boundp state 'amplitudes)
+          (not (null (slot-value state 'amplitudes))))
+     (assert (<= num-qubits (wavefunction-qubits (amplitudes state)))
+             ()
+             state
+             (wavefunction-qubits (amplitudes state))
+             num-qubits))
+    (t
+     (setf
+      ;; Initialize the AMPLITUDES and TRIAL-AMPLITUDES to an empty
+      ;; array of the correct size.
+      (amplitudes state) (make-lisp-cflonum-vector (expt 2 num-qubits)))))
+  (setf
+     (%trial-amplitudes state) (make-lisp-cflonum-vector (expt 2 num-qubits))
+      ;; Save a pointer to the originally provided memory.
+      (slot-value state 'original-amplitudes) (amplitudes state)))
+
+(defmethod num-qubits ((state pure-state))
+  ;; Return the number of qubits that the STATE represents.
+  (quil:ilog2 (length (amplitudes state))))
+
+(defmethod set-to-zero-state ((state pure-state))
+  ;; Return the STATE to the ground state.
+  (bring-to-zero-state (amplitudes state)))
+
+(defmethod requires-swapping-amps-p ((state pure-state))
+  ;; Does the STATE require swapping of internal pointers? This
+  ;; function is used after stochastic state evolution occurs to a
+  ;; PURE-STATE.
+  (and (not (eq (amplitudes state) (original-amplitudes state)))
+       #+sbcl (eq ':foreign (sb-introspect:allocation-information
+                             (original-amplitudes state)))))
+
+(defmethod swap-internal-amplitude-pointers ((state pure-state))
+  ;; Copy the correct amplitudes into place.
+  (copy-wavefunction (amplitudes state) (original-amplitudes state))
+  ;; Get the pointer back in home position. We want to swap them,
+  ;; not overwrite, because we want the scratch memory to be intact.
+  (rotatef (amplitudes state) (%trial-amplitudes state)))
+
+
+;;; The DENSITY-MATRIX-STATE is a representation of a density matrix
+;;; quantum state that can be operated upon by a QVM.
+(defclass density-matrix-state (quantum-system-state)
+  ((amplitudes
```

Comment:
I'd suggest renaming this to something like `elements-vector`. The only reason I was using `amplitudes` before for the vector of matrix elements was because I was piggybacking on the pure state representation. But they are not "amplitudes" in the usual sense. Since you're doing the right thing this time around, this is an opportunity to fix this idiosyncratic name.

Reply:
For this to work though, it would probably require lifting some generics (like `reset-quantum-state) from having a method defined on `base-qvm` to instead having methods defined on the subclasses (e.g. on `pure-state-qvm`). Likewise `amplitudes` accessor and `setf` stuff would have to be updated accordingly.

Reply:
This sounds good and makes sense to me -- @stylewarning also suggested a similar change. Would it be worth renaming the `qvm`-level `amplitudes` method to something like `state-elements`, so that it is more representative of the state data of both PURE-STATE and DENSITY-MATRIX-STATE types?

Reply:
^^ this actually looks like it would be a pretty major change (involving changes to `dqv` and the `qvm-app`) so maybe its best as a follow up issue.